pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -629,53 +641,48 @@ impl<T: Config> Module<T> {
-        let seconds_elapsed = T::BillingFrequency::get() * 6;
-
-        // Calculate the cost for a contract, can be any of:
-        // - NodeContract
-        // - RentContract
-        // - NameContract
-        let total_cost = Self::calculate_contract_cost(contract, &pricing_policy, seconds_elapsed)?;
-
-        // If cost is 0, reinsert to be billed at next interval
-        if total_cost == 0 {
-            return Ok(());
-        }
-
-        let total_cost_tft_64 = Self::calculate_cost_in_tft_from_musd(total_cost)?;
-
         if !pallet_tfgrid::Twins::<T>::contains_key(contract.twin_id) {
             return Err(DispatchError::from(Error::<T>::TwinNotExists));
         }
         let twin = pallet_tfgrid::Twins::<T>::get(contract.twin_id);
-        let balance: BalanceOf<T> = <T as Config>::Currency::free_balance(&twin.account_id);
 
-        // Calculate the amount due and discount received based on the total_cost amount due
-        let (amount_due, discount_received) =
-            Self::calculate_discount(total_cost_tft_64, balance, certification_type);
+        let mut balance: BalanceOf<T> = <T as Config>::Currency::free_balance(&twin.account_id);
 
-        // Convert amount due to u128
-        let amount_due_as_u128: u128 = amount_due.saturated_into::<u128>();
-        // Get current TFT price
+        let (mut amount_due, discount_received) =
+            Self::calculate_contract_cost_tft(contract, balance)?;
 
-        // if the total amount due exceeds the twin's balance, decomission contract
-        // but first drain the account with the amount equal to the balance of that twin
-        let mut amount_due: BalanceOf<T> = BalanceOf::<T>::saturated_from(amount_due_as_u128);
+        if amount_due == BalanceOf::<T>::saturated_from(0 as u128) {
+            return Ok(());
+        };
 
+        // If the contract is in cancel state, which means that this is the last billing cycle
+        // remove the lock on the funds so they become free and we can bill these if needed
+        if matches!(
+            contract.state,
+            types::ContractState::Deleted(types::Cause::CanceledByUser)
+        ) {
+            <T as Config>::Currency::unreserve(&twin.account_id, amount_due);
+            balance = <T as Config>::Currency::free_balance(&twin.account_id);
+        }
+
+        // if the total amount due exceeds the twin's balance, first try to unreserve the balance
+        // if the total balance after unreserving can't still cover the amount due, decomission the contract
         let mut decomission = false;
         if amount_due >= balance {
-            debug::info!(
-                "decomissioning contract because balance on twin account is lower than amount due"
-            );
-            amount_due = balance;
-            decomission = true;
+            <T as Config>::Currency::unreserve(&twin.account_id, amount_due);
+            balance = <T as Config>::Currency::free_balance(&twin.account_id);
+
```

Comment:
Personally I think we should always decomission here. Fact is that the user did not have enough tokens (and so will likely default in the next billing cycle anyway), and so we should decomission here. While he __technically__ paid for this cycle, allowing him another cycle kind of removes the utility of the lock as he can get an hour of usage for free then without paying, which is the original issue

Reply:
Yes indeed! Woops